cpp/unreal_plugins/SpEngine/Source/SpEngine/LegacyService.h

```diff
@@ -0,0 +1,337 @@
+
+                // if we're at this line of code and OpenLevel is already pending, it means we failed
+                SP_ASSERT(!open_level_pending_);
+
+                UGameplayStatics::OpenLevel(world, Unreal::toFName(desired_level_name));
+                open_level_pending_ = true;
+
+            } else {
+                open_level_pending_ = false;
+
+                // we expect worldCleanupHandler(...) to be called before a new world is created
+                SP_ASSERT(!world_);
+
+                // cache local reference to the UWorld
+                world_ = world;
+
+                // We need to defer initializing this handler until after we have a valid world_ pointer,
+                // and we defer the rest of our initialization code until the OnWorldBeginPlay event.
+                world_begin_play_handle_ = world_->OnWorldBeginPlay.AddRaw(this, &LegacyService::worldBeginPlayEventHandler);
+            }
+        }
+    }
+
+    void worldCleanupHandler(UWorld* world, bool session_ended, bool cleanup_resources)
+    {
+        SP_LOG_CURRENT_FUNCTION();
+        SP_ASSERT(world);
+
+        // We only need to perform any additional steps if the world being cleaned up is the world we cached in our world_ member variable.
+        if (world == world_) {
+
+            // The worldCleanupHandler(...) function is called for all worlds, but some local state (such as rpc_server_ and agent_)
+            // is initialized only when worldBeginPlayEventHandler(...) is called for a particular world. So we check if worldBeginPlayEventHandler(...)
+            // has been executed.
+            if (has_world_begin_play_executed_) {
+                has_world_begin_play_executed_ = false;
+
+                SP_ASSERT(nav_mesh_);
+                nav_mesh_->cleanUpObjectReferences();
+                nav_mesh_ = nullptr;
+
+                SP_ASSERT(task_);
+                task_->cleanUpObjectReferences();
+                task_ = nullptr;
+
+                SP_ASSERT(agent_);
+                agent_->cleanUpObjectReferences();
+                agent_ = nullptr;
+            }
+
+            // remove event handlers bound to this world before world gets cleaned up
+            world_->OnWorldBeginPlay.Remove(world_begin_play_handle_);
+            world_begin_play_handle_.Reset();
+
+            // clear cached world_ pointer
+            world_ = nullptr;
+        }
+    }
+
+    void worldBeginPlayEventHandler()
+    {
+        SP_LOG_CURRENT_FUNCTION();
+        SP_ASSERT(world_);
+
+        // execute optional console commands from python client
+        if (Config::isInitialized()) {
+            for (auto& command : Config::get<std::vector<std::string>>("SP_ENGINE.LEGACY_SERVICE.CUSTOM_UNREAL_CONSOLE_COMMANDS")) {
+                GEngine->Exec(world_, *Unreal::toFString(command));
+            }
+        }
+
+        // set physics parameters
+        UPhysicsSettings* physics_settings = UPhysicsSettings::Get();
```

Comment:
In a future PR we should move this engine initialization code to `EngineActor` so we get consistent behavior in the editor and the standalone game, while also being easy to tweak interactively in the editor, and being easy to tweak via a config file. Each engine initialization value should be backed by a `UPROPERTY` on `EngineActor`, or maybe we can just have a single `UPhysicsSettings*` object as a `UPROPERTY` that would handle everything.

Reply:
Exposing `UPhysicsSettings` as a `UPROPERTY` sounds like a good idea. It's much cleaner this way. Definitely worth a shot.